src/Cardano/Wallet/Api/Types/Percentage.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Cardano.Wallet.Api.Types.Percentage
+    ( Percentage
+    , PercentageError (..)
+    , mkPercentage
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Api.JSON
+    ( defaultRecordTypeOptions )
+import Data.Aeson
+    ( FromJSON (..), ToJSON (..), genericParseJSON, genericToJSON )
+import Data.Text
+    ( Text )
+import GHC.Generics
+    ( Generic )
+
+newtype Percentage = Percentage
+    { getPercentage :: Int
+    } deriving (Eq, Generic, Ord, Show)
+
+data PercentageError
+    = PercentageOutOfBoundsError
+    | PercentageUnitInvalidError
+    deriving Show
+
+instance Bounded Percentage where
+    minBound = Percentage 0
+    maxBound = Percentage 100
+
+instance Enum Percentage where
+    toEnum = unsafeMkPercentage
+    fromEnum = getPercentage
+
+instance FromJSON Percentage where
+    parseJSON x = either (fail . show) pure . validate =<< parseJSON x
+instance ToJSON Percentage where
+    toJSON = toJSON . unvalidate
+
+mkPercentage :: Integral i => i -> Either PercentageError Percentage
+mkPercentage i
+    | j < minBound = Left PercentageOutOfBoundsError
+    | j > maxBound = Left PercentageOutOfBoundsError
+    | otherwise = pure j
+  where
+    j = Percentage $ fromIntegral i
+
+unsafeMkPercentage :: Integral i => i -> Percentage
+unsafeMkPercentage = either (error . show) id . mkPercentage
+
+data UnvalidatedPercentage = UnvalidatedPercentage
+    { _quantity :: Int
+    , _unit :: Text
```

Comment:
I know it matches the `swagger.yaml` specification, but this `unit: "percent"` thing is hokey. There are no other units than percent. I reckon the property name should just be changed. E.g. `progress` → `progressPercent`.

Reply:
You're right that there are no units other than `percent`.

I used the `UnvalidatedPercentage` type in order to make JSON encoding easier, because it's very close to the JSON representation. As part of validation during deserialization, we transform an `UnvalidatedPercentage` to a `Percentage`, which is the idiomatic Haskell representation.

Notice that we only export `Percentage` from the module, and not `UnvalidatedPercentage` (values of which only exist transiently).